python/deltalake/write_table.py

```diff
@@ -0,0 +1,546 @@
+        file_options,
+        filesystem,
+        max_partitions,
+        partition_by,
+):
+    max_open_files: int = 1024
+    max_rows_per_file: int = 10 * 1024 * 1024
+    min_rows_per_group: int = 64 * 1024
+    max_rows_per_group: int = 128 * 1024
+    data = batch_ref
+    add_actions: List[AddAction] = []
+    if partition_by != None:
+        partition_schema = pa.schema(
+            filter(lambda x: x[0] in partition_by, [(name, ntype) for name, ntype in zip(schema.names, schema.types)]))
+        partitioning = ds.partitioning(partition_schema, flavor="hive")
+    else:
+        partitioning = None
+
+    def visitor(written_file: Any) -> None:
+        # TODO Fix the partitioning path
+        path, partition_values = get_partitions_from_path(written_file.path)
+        stats = get_file_stats_from_metadata(written_file.metadata)
+        # PyArrow added support for written_file.size in 9.0.0
+        if PYARROW_MAJOR_VERSION >= 9:
+            size = written_file.size
+        else:
+            size = filesystem.get_file_info([path])[0].size  # type: ignore
+
+        add_actions.append(
+            AddAction(
+                path,
+                size,
+                partition_values,
+                int(datetime.now().timestamp() * 1000),
+                True,
+                json.dumps(stats, cls=DeltaJSONEncoder),
+            )
+        )
+    ds.write_dataset(
+        data,
+        base_dir="/",
+        basename_template=f"{current_version + 1}-{uuid.uuid4()}-{{i}}.parquet",
+        format="parquet",
+        partitioning=partitioning,
+        # It will not accept a schema if using a RBR
+        schema=schema,
+        file_visitor=visitor,
+        existing_data_behavior="overwrite_or_ignore",
+        file_options=file_options,
+        max_open_files=max_open_files,
+        max_rows_per_file=max_rows_per_file,
+        min_rows_per_group=min_rows_per_group,
+        max_rows_per_group=max_rows_per_group,
+        filesystem=filesystem,
+        max_partitions=max_partitions,
+    )
+    return add_actions
+
+
+def write_deltalake_ray(
+        table_or_uri: Union[str, Path, DeltaTable],
+        data: Union["ray.data.dataset.Dataset"],
+        *,
+        schema: Optional[pa.Schema] = None,
+        partition_by: Optional[Union[List[str], str]] = None,
+        filesystem: Optional[pa_fs.FileSystem] = None,
+        mode: Literal["error", "append", "overwrite", "ignore"] = "error",
+        file_options: Optional[ds.ParquetFileWriteOptions] = None,
+        max_partitions: Optional[int] = None,
+        name: Optional[str] = None,
+        description: Optional[str] = None,
+        configuration: Optional[Mapping[str, Optional[str]]] = None,
+        overwrite_schema: bool = False,
+        storage_options: Optional[Dict[str, str]] = None,
+        partition_filters: Optional[List[Tuple[str, str, Any]]] = None,
+) -> None:
+    import os
+    os.environ["AWS_S3_ALLOW_UNSAFE_RENAME"] = "true"
```

Comment:
defaulting to usafe behaivour is probably something we should avoid :)

Reply:
cleaned